test/Microsoft.Extensions.Internal.Test/TypeNameHelperTest.cs

```diff
@@ -128,6 +142,66 @@ public void Can_pretty_print_array_name(string expected, Type type, bool fullNam
             Assert.Equal(expected, TypeNameHelper.GetTypeDisplayName(type, fullName));
         }
 
+        public static IEnumerable<object[]> GetOpenGenericsTestData()
+        {
+            yield return new object[] { "List<>", typeof(List<>), false };
+            yield return new object[] { "Dictionary<,>", typeof(Dictionary<,>), false };
+            yield return new object[] { "System.Collections.Generic.List<>", typeof(List<>) };
+            yield return new object[] { "System.Collections.Generic.Dictionary<,>",typeof(Dictionary<,>) };
+
+            yield return new object[]
+            {
+                "Microsoft.Extensions.Internal.TypeNameHelperTest+Level1<>+Level2<>+Level3<>",
+                typeof(Level1<>.Level2<>.Level3<>)
+            };
+
+            yield return new object[]
+            {
+                "Microsoft.Extensions.Internal.TypeNameHelperTest+OuterGeneric<>+InnerNonGeneric+InnerGeneric<,>+InnerGenericLeafNode<>",
+                typeof(OuterGeneric<>.InnerNonGeneric.InnerGeneric<,>.InnerGenericLeafNode<>)
+            };
+
+            var openDictionaryType = typeof(Dictionary<,>);
+            var genArgsDictionary = openDictionaryType.GetGenericArguments();
+            genArgsDictionary[0] = typeof(B<>);
+            var closedDictionaryType = openDictionaryType.MakeGenericType(genArgsDictionary);
+
+            yield return new object[]
+            {
+                "System.Collections.Generic.Dictionary<Microsoft.Extensions.Internal.TypeNameHelperTest+B<>,>",
+                closedDictionaryType
+            };
+
+            var openLevelType = typeof(Level1<>.Level2<>.Level3<>);
+            var genArgsLevel = openLevelType.GetGenericArguments();
+            genArgsLevel[1] = typeof(string);
```

Comment:
If this types are not constructible without reflection then leave it. We don't need tests for that. Test may verify the pipeline but no point of writing tests which are impossible without ugly hacks.

Reply:
Those types totally possible without reflection, using inheritance. It's just easier to construct them it that way than creating class hierarchy for each one.